Adding ArcAssay/ArcStudy API

Freymaurer opened the pull request:
- Hit >300 unit tests 🎉 ✅ 
- JS tests not yet run ❌

Freymaurer commented:
@HLWeil please have a look!

HLWeil commented on src/ISA/ISA/ArcTypes/ArcAssay.fs:
```diff
@@ -2,6 +2,64 @@
 
+module ArcAssayAux =
+
+    let getNextAutoGeneratedTableName (existingNames: seq<string>) =
```
I would enforce users to set table names themselves when using this API.

HLWeil commented on src/ISA/ISA/ArcTypes/ArcTable.fs:
```diff
@@ -400,4 +422,17 @@ type ArcTable =
+        | _ -> false
+
+    // it's good practice to ensure that this behaves using the same fields as Equals does:
+    override this.GetHashCode () = this.GetHashCode()
```
rip

HLWeil commented on src/ISA/ISA/ArcTypes/ArcAssay.fs:
```diff
@@ -24,26 +82,349 @@ type ArcAssay =
+
+    // - Table API - //
+    // remark should this return ArcTable?
+    member this.AddTable(?table:ArcTable, ?index: int) = 
```
```fsharp
member this.AddTable(table:ArcTable, ?index: int) : unit

member this.InitTable(tableName : string, ?index : int) : ArcTable
```

HLWeil replied:
Maybe something like this?

HLWeil commented on src/ISA/ISA/ArcTypes/ArcAssay.fs:
```diff
@@ -24,26 +82,349 @@ type ArcAssay =
+
+    // - Table API - //
+    member this.GetTable(name: string) : ArcTable =
+        tryByTableName name this.Tables
```
rename

HLWeil commented on src/ISA/ISA/ArcTypes/ArcAssay.fs:
```diff
@@ -24,26 +82,349 @@ type ArcAssay =
+    static member getTableAt(index:int) : ArcAssay -> ArcTable =
+        fun (assay:ArcAssay) ->
+            let newAssay = assay.Copy()
+            newAssay.GetTableAt(index)
```
- add comment to make clear that the user receives a copied object
- copy table, not assay

HLWeil reviewed (changes requested):
<3

Freymaurer commented:
Renamed "Set" CRUD api functions to "Updated"; `SetColumn` --> `UpdateColumn`
Renamed `UpdateTableAt` functions on ArcAssay and ArcStudy to `MapTableAt`

```fsharp
member this.MapTableAt(index: int, updateFun: ArcTable -> unit) = ArcTables(this.Tables).MapTableAt(index, updateFun)
```

HLWeil commented:
Great work! 

I will merge this, so I can start using these functions in the FsSpreadsheet IO. If we decide on additional changes feel free to open up a new PR.
